### PR TITLE
Dispatch VALUE_CHANGE on `amp-bind` value changes

### PR DESCRIPTION
**Pull request by xrav3nz**
This introduces a new `AmpEvents.VALUE_CHANGE` event. It is dispatched
when `amp-bind` mutates form input values, specifically:
`<input [value]>`, `<textarea [text]>`, `<input [checked]>`, and
`<option [selected]>`.

A custom event is chosen over a mutation observer, because binding
`[text]` on `textarea` does not change the DOM (and it's expected), thus
a mutation observer won't pick up the change. The same applies to
`input` elements, when we eventually have two separate bindings for
`[value]` and `[defaultValue]`.

Part of #22534.

/cc @GoTcWang @choumx thanks!

### CODE REVIEW

File: extensions/amp-bind/0.1/test/integration/test-bind-impl.js [hunk 1]
```diff
@@ -1138,6 +1138,109 @@ describe
           yield onBindReadyAndSetState(env, bind, {foo: 'bar'});
           expect(foo.textContent).to.not.equal('bar');
         });
+
+        describe('AmpEvents.VALUE_CHANGE dispatches', () => {
```
Comment (GoTcWang) [inline extensions/amp-bind/0.1/test/integration/test-bind-impl.js]: nit: remove "AmpEvents.VALUE_CHANGE"?

It sounds more like a sentence "AmpEvents.VALUE_CHANGE" and 'should dispatch on <input [value]> changes'

File: src/amp-events.js [hunk 2]
```diff
@@ -21,6 +21,7 @@
 export const AmpEvents = {
   BUILT: 'amp:built',
   DOM_UPDATE: 'amp:dom-update',
+  VALUE_CHANGE: 'amp:value-change',
```
Comment (GoTcWang) [inline src/amp-events.js]: How do we plan on using this event? Is the situation in this PR the only situation we would like to use it?

The reason why I am asking is this feels more like "amp-form input value change" event only from this PR.

File: extensions/amp-bind/0.1/test/integration/test-bind-impl.js (hunk 1 above)
Comment (xrav3nz) [reply extensions/amp-bind/0.1/test/integration/test-bind-impl.js]: Made the test name more elaborate so there is no confusion "should dispatch VALUE_CHANGE on ..."

File: src/amp-events.js (hunk 2 above)
Comment (xrav3nz) [reply src/amp-events.js]: I think it's best if this event is only dispatched for form field value changes.

`FORM_VALUE_CHANGE: 'amp:form-value-change'`? Not sure if "input" would be ambiguous since `<input>` element vs general user input fields. I don't feel strongly about either tho.

Comment (GoTcWang) [reply src/amp-events.js]: FORM_VALUE_CHANGE SG

File: extensions/amp-bind/0.1/bind-impl.js [hunk 3]
```diff
@@ -1646,3 +1651,38 @@ export class Bind {
...
+  switch (property) {
+    case 'checked':
+    case 'value':
+      return tagName === 'INPUT';
+    case 'selected':
+      return tagName === 'OPTION';
+    case 'text':
+      return tagName === 'TEXTAREA';
+    default:
+      return false;
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Ah, I didn't realize there were so many cases. Might as well add `[disabled]` for form elements here too then. 

Then semantically this function becomes "any change, observable or not, that affects form data" instead of "only unobservable changes".

**Comment by xrav3nz** [comment]
Updated the PR to:
- Dispatch the new event at the parent `<select>` when `<option [selected]>` changes, similar to the native `InputEvent`
- Dispatch the new event on `disabled` attribute changes.
- Dispatch the new event on `type` attribute changes - the motivation is to catch `type=hidden` changes, but I am not sure if it is common to hide with `type` rather than a CSS class.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 4]
```diff
@@ -1646,3 +1656,49 @@ export class Bind {
     }
   }
 }
+
+/** @private @const {!Object<string, boolean>} */
+const CAN_BE_DISABLED = {
```
Comment (GoTcWang) [inline extensions/amp-bind/0.1/bind-impl.js]: CAN_BE_DISABLED_FORM_FIELD?

File: extensions/amp-bind/0.1/bind-impl.js [hunk 5]
```diff
@@ -1646,3 +1656,49 @@ export class Bind {
...
+
+/**
+ * Returns whether the element's property affects the value of a form field.
+ * @param {string} tagName
+ * @param {string} property
+ * @return {boolean}
+ */
+function doesPropertyAffectFormValue(tagName, property) {
+  switch (property) {
+    case 'disabled':
```
Comment (GoTcWang) [inline extensions/amp-bind/0.1/bind-impl.js]: Should we consider [hidden] attribute as well in this case then? Feel free to left a TODO and resolve in another PR.

Note that `hidden` attribute can be mutated on the ancestry of a form field.

File: extensions/amp-bind/0.1/test/integration/test-bind-impl.js [hunk 6]
```diff
@@ -1138,6 +1138,151 @@ describe
...
+              'input'
+            );
+            const spy = sandbox.spy(element, 'dispatchEvent');
+
+            return onBindReadyAndSetState(env, bind, {foo: 'disabled'}).then(
+              () => {
+                expect(spy).to.have.been.calledOnce;
+                expect(spy).calledWithMatch({
+                  type: AmpEvents.FORM_VALUE_CHANGE,
+                  bubbles: true,
```
Comment (GoTcWang) [inline extensions/amp-bind/0.1/test/integration/test-bind-impl.js]: I feel this whole request: 
{
                  type: AmpEvents.FORM_VALUE_CHANGE,
                  bubbles: true,
                }
can be extracted to be a constant, so you don't need to have it across all tests.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 5 above)
Comment (xrav3nz) [reply extensions/amp-bind/0.1/bind-impl.js]: Ugh right. So there are:

- `disabled` for form fields and fieldsets
- `hidden` for all essentially all elements
- `type` for `<input>`

I am still leaning towards using a mutation observer on the form for these, and leave `AmpEvents.FORM_VALUE_CHANGE` as close to the native `InputEvent` as possible. (It's also semantically cleaner as Will mentioned)

Thoughts? @choumx @GoTcWang

Comment (GoTcWang) [reply extensions/amp-bind/0.1/bind-impl.js]: Yup that also works.
I'll left it to Will to make the call.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 7]
```diff
@@ -1138,16 +1138,26 @@ export class Bind {
...
           if (property == 'width') {
             width = isFiniteNumber(newValue) ? Number(newValue) : width;
           } else if (property == 'height') {
             height = isFiniteNumber(newValue) ? Number(newValue) : height;
           }
         }
+
+        if (isPropertyAFormValue(element.tagName, property)) {
+          const dispatchAt =
+            element.tagName === 'OPTION' ? element.closest('SELECT') : element;
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Out of curiosity, why is this necessary? Also use `closestAncestorElementBySelector()` helper in dom.js in case `Element.closest` is not available.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 8]
```diff
@@ -1138,16 +1138,26 @@ export class Bind {
...
+
+          if (dispatchAt) {
+            dispatchAmpFormValueChangeEvent(this.localWin_, dispatchAt);
+          }
+        }
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Nit: Can we keep everything in a single class function? It would also save you from checking `tagName` twice.

```js
this.dispatchFormValueChangeEventIfNecessary_(element, property);
```

File: extensions/amp-bind/0.1/bind-impl.js (hunk 5 above)
Comment (choumx) [reply extensions/amp-bind/0.1/bind-impl.js]: SGTM

File: extensions/amp-bind/0.1/bind-impl.js (hunk 8 above)
Comment (xrav3nz) [reply extensions/amp-bind/0.1/bind-impl.js]: To only check `tagName` once, the `switch` gets a bit intense with the nested `if`s and early returns: https://gist.github.com/xrav3nz/6fd7b048080aedbaa1e87c418b4193e4 Are you okay with this?

File: extensions/amp-bind/0.1/bind-impl.js (hunk 7 above)
Comment (xrav3nz) [reply extensions/amp-bind/0.1/bind-impl.js]: Thanks! The native `InputEvent` is also dispatched at the parent `<select>` when the selected `<option>` changes. I think it makes sense, since it is the value of the `<select>` element that is changed.

I just added a comment explaining this!

File: extensions/amp-bind/0.1/bind-impl.js (hunk 8 above)
Comment (choumx) [reply extensions/amp-bind/0.1/bind-impl.js]: True, probably better as is. Speaking of switch statements, a lookup map might be more readable. 

Optional:
```js
/** @const {!Object<string, !Array<string>>} */
const FORM_VALUE_PROPERTIES = {
  'INPUT': ['checked', 'value'],
  'OPTION': ['selected'],
  'TEXTAREA': ['text'],
};

dispatchFormValueChangeEventIfNecessary_(element, property) {
  const props = FORM_VALUE_PROPERTIES[element.tagName];
  if (props && props.includes(property)) {
    ...
  }
}
```

Comment (xrav3nz) [reply extensions/amp-bind/0.1/bind-impl.js]: I think I'm gonna to leave it as is. Thanks again for the reviews! 💪

**Comment by choumx** [comment]
Nice work, thanks for contributing.
